## 1. Import Required Libraries

In [17]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from tensorflow.keras.datasets import imdb
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, SimpleRNN, Dense, Dropout
from tensorflow.keras.callbacks import EarlyStopping

from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
import seaborn as sns

## 2. Load IMDb Dataset

The IMDb dataset contains 50,000 movie reviews.

25,000 Training Reviews
25,000 Testing Reviews

Each review is already encoded into integers representing words.

In [18]:
vocab_size = 10000

(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=vocab_size)

print("Training Samples:", len(X_train))
print("Testing Samples :", len(X_test))

Training Samples: 25000
Testing Samples : 25000


## 3. Pad the Sequences

Movie reviews have different lengths.

Pad them to equal length.

In [19]:
max_length = 200

X_train = pad_sequences(X_train,
                        maxlen=max_length,
                        padding='post')

X_test = pad_sequences(X_test,
                       maxlen=max_length,
                       padding='post')

## 4. Build the RNN Model

In [20]:
model = Sequential()

model.add(Embedding(input_dim=vocab_size,
                    output_dim=128,
                    input_length=max_length))

model.add(SimpleRNN(64))

model.add(Dropout(0.5))

model.add(Dense(1, activation='sigmoid'))

model.summary()

C:\Users\LENOVO\AppData\Roaming\Python\Python313\site-packages\keras\src\layers\core\embedding.py:103: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ embedding_1 (Embedding)              │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ simple_rnn_1 (SimpleRNN)             │ ?                           │     0 (unbuilt) │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dropout_1 (Dropout)                  │ ?                           │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ ?                           │     0 (unbuilt) │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

## 5. Compile the Model

In [21]:
model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

## 6. Train the Model

In [22]:
early_stop = EarlyStopping(monitor='val_loss',
                           patience=2,
                           restore_best_weights=True)

history = model.fit(
    X_train,
    y_train,
    validation_split=0.2,
    epochs=10,
    batch_size=128,
   
)

Epoch 1/10
157/157 ━━━━━━━━━━━━━━━━━━━━ 13s 71ms/step - accuracy: 0.5173 - loss: 0.6934 - val_accuracy: 0.5460 - val_loss: 0.6828
Epoch 2/10
157/157 ━━━━━━━━━━━━━━━━━━━━ 11s 69ms/step - accuracy: 0.6194 - loss: 0.6354 - val_accuracy: 0.5366 - val_loss: 0.6850
Epoch 3/10
157/157 ━━━━━━━━━━━━━━━━━━━━ 11s 67ms/step - accuracy: 0.6928 - loss: 0.5340 - val_accuracy: 0.5486 - val_loss: 0.7118
Epoch 4/10
157/157 ━━━━━━━━━━━━━━━━━━━━ 10s 66ms/step - accuracy: 0.7452 - loss: 0.4311 - val_accuracy: 0.5334 - val_loss: 0.7673
Epoch 5/10
157/157 ━━━━━━━━━━━━━━━━━━━━ 10s 66ms/step - accuracy: 0.7750 - loss: 0.3847 - val_accuracy: 0.5396 - val_loss: 0.8505
Epoch 6/10
157/157 ━━━━━━━━━━━━━━━━━━━━ 11s 68ms/step - accuracy: 0.7929 - loss: 0.3571 - val_accuracy: 0.5440 - val_loss: 0.8204
Epoch 7/10
157/157 ━━━━━━━━━━━━━━━━━━━━ 11s 67ms/step - accuracy: 0.8194 - loss: 0.3224 - val_accuracy: 0.5314 - val_loss: 0.9249
Epoch 8/10
157/157 ━━━━━━━━━━━━━━━━━━━━ 11s 67ms/step - accuracy: 0.8280 - loss: 0.3189 - 

## 7. Evaluate the Model

In [23]:
loss, accuracy = model.evaluate(X_test, y_test)

print("Test Accuracy:", accuracy)
print("Test Loss:", loss)

782/782 ━━━━━━━━━━━━━━━━━━━━ 7s 8ms/step - accuracy: 0.5385 - loss: 1.0044
Test Accuracy: 0.5385199785232544
Test Loss: 1.0043857097625732


## 8. Make Predictions

In [ ]:
y_pred_prob = model.predict(X_test)

y_pred = (y_pred_prob > 0.5).astype(int)

241/782 ━━━━━━━━━━━━━━━━━━━━ 8s 15ms/step

## 9. Classification Report

In [ ]:
print(classification_report(y_test, y_pred))

## 10. Confusion Matrix

In [ ]:
cm = confusion_matrix(y_test, y_pred)

sns.heatmap(cm,
            annot=True,
            fmt='d',
            cmap='Blues')

plt.xlabel("Predicted")
plt.ylabel("Actual")
plt.title("Confusion Matrix")
plt.show()

## 11. Plot Accuracy

In [ ]:
plt.plot(history.history['accuracy'], label='Training Accuracy')
plt.plot(history.history['val_accuracy'], label='Validation Accuracy')

plt.xlabel("Epoch")
plt.ylabel("Accuracy")
plt.title("Training vs Validation Accuracy")
plt.legend()
plt.show()

## 12. Plot Loss

In [ ]:
plt.plot(history.history['loss'], label='Training Loss')
plt.plot(history.history['val_loss'], label='Validation Loss')

plt.xlabel("Epoch")
plt.ylabel("Loss")
plt.title("Training vs Validation Loss")
plt.legend()
plt.show()

In [ ]:
model.save("sentiment_rnn_model.keras")

print("Model saved successfully!")

In [ ]:
import pickle

word_index = imdb.get_word_index()

with open("word_index.pkl", "wb") as f:
    pickle.dump(word_index, f)

print("Word Index Saved Successfully!")

# Conclusion
The Sentiment Analysis System using SimpleRNN successfully classified IMDb movie reviews into positive and negative sentiments. The model achieved approximately 78.66% test accuracy, with precision, recall, and F1-score around 0.79. Although some overfitting was observed, the model demonstrates that RNNs can learn sequential word patterns for sentiment classification.

Step 16. Streamlit Application Development
Overview
To make the sentiment analysis model accessible to end users, a Streamlit web application was developed. The application provides an interactive interface where users can enter movie reviews and receive real-time sentiment predictions generated by the trained SimpleRNN model.

The application loads the saved deep learning model and word index, preprocesses the input text, predicts the sentiment, and displays the result along with a confidence score and interactive visualizations.

Deployment Workflow
The following steps were performed to build the Streamlit application:

Created a new app.py file.
Loaded the trained SimpleRNN model (.keras).
Loaded the saved word_index.pkl file.
Developed text preprocessing functions for user input.
Converted input reviews into integer sequences.
Applied sequence padding to match the training input length.
Predicted the sentiment using the trained model.
Displayed the predicted sentiment and confidence score.
Designed a modern Glassmorphism-based user interface using custom CSS.
Added a sentiment confidence progress bar.
Added a sentiment gauge for visual interpretation.
Included clickable sample movie reviews for quick testing.
Displayed model information and prediction metrics.
Prepared the application for deployment using Streamlit.
Application Features
Interactive Movie Review Input
Automatic Text Cleaning and Preprocessing
Real-Time Sentiment Prediction
Confidence Score Visualization
Sentiment Gauge
Glassmorphism User Interface
Clickable Sample Reviews
Responsive Dashboard Layout
Files Required for Deployment
app.py
sentiment_rnn_model.keras
word_index.pkl
requirements.txt
README.md
Running the Application
Install the required libraries:

pip install -r requirements.txt
Run the Streamlit application:

streamlit run app.py
The application will launch in the default web browser, allowing users to enter movie reviews and instantly obtain sentiment predictions.

Final Conclusion
In this project, a Sentiment Analysis System was successfully developed using a Simple Recurrent Neural Network (SimpleRNN) to classify IMDb movie reviews as positive or negative. The model was trained on the IMDb dataset, evaluated using standard classification metrics, and achieved satisfactory performance on unseen data.

Furthermore, the trained model was deployed as an interactive Streamlit web application, enabling users to perform real-time sentiment analysis through a modern and user-friendly interface. This project demonstrates the complete workflow of an end-to-end deep learning application, from data preprocessing and model development to evaluation and deployment.